dmscripts/get_log_events.py

```diff
@@ -0,0 +1,112 @@
+        self.client = session.client('logs')
+
+    def filter_streams_in_interval(self, streams, earliest_epoch, latest_epoch):
+        """
+        Parses all the stream object of boto's describe_log_streams,
+        filtering out those without any entries in the desired interval
+        """
+        filtered_streams = []
+        for stream in streams:
+            # covert javascript/AWS milliseconds to python/ISO seconds
+            stream['firstEventTimestamp'] /= 1000
+            stream['lastEventTimestamp'] /= 1000
+            # append the UUID of the log in the interval
+            if earliest_epoch < stream['lastEventTimestamp'] and stream['firstEventTimestamp'] < latest_epoch:
+                filtered_streams.append(stream['logStreamName'])
+        return filtered_streams
+
+    def describe_log_streams(self, earliest_epoch, latest_epoch):
+        streams = []
+        # fetch first log stream
+        response = self.client.describe_log_streams(
+            logGroupName=self.group,
+            orderBy='LastEventTime',
+            descending=True
+        )
+        # check epoch range and add stream logStreamName
+        streams.extend(self.filter_streams_in_interval(response['logStreams'], earliest_epoch, latest_epoch))
+        while len(response['logStreams']) > 0 and 'nextToken' in response.keys():
+            response = self.client.describe_log_streams(
+                logGroupName=self.group,
+                orderBy='LastEventTime',
+                descending=True,
+                nextToken=response['nextToken']
+            )
+            # check epoch range and add stream logStreamName
+            streams.extend(self.filter_streams_in_interval(response['logStreams'], earliest_epoch, latest_epoch))
+        return streams
+
+    def get_log_events(self, stream, directory):
+        """
+        see example use in ./scripts/oneoff/get-log-events.py
+        """
+        print(f'Fetching {stream} stream.')
+        counter = 0
+        directory = os.path.join(directory, self.group, stream)
+        os.makedirs(directory, exist_ok=True)
+        # fetch the logs first page
+        response = self.client.get_log_events(
+            logGroupName=self.group,
+            logStreamName=f'{stream}',
+            startFromHead=True
+        )
+        with open(f'{directory}/{counter}.json', 'w') as f:
+            json.dump(response, f)
+        # iterate to fetch the rest of the pages
+        while len(response['events']) > 0:
+            counter += 1
+            response = self.client.get_log_events(
+                logGroupName=self.group,
+                logStreamName=stream,
+                startFromHead=True,
+                nextToken=response['nextForwardToken']
+            )
+            with open(f'{directory}/{counter}.json', 'w') as f:
+                json.dump(response, f)
+        return counter + 1
+
+    def get_log_event_in_epoch_range(self, directory, earliest_epoch, latest_epoch):
```

Comment:
Could we supply the extra args to the main script and use it conditionally if those are present? Or do we just want to keep it as an option in a Python shell?

Reply:
Yes, but I'd rather have a new script using the same class, instead of working with optional argument in docopt.